### PR TITLE
Changed Security Structure 

### PR DESCRIPTION
**Pull request by bennchong**
From previously of having a separate SecurityBox and SecurityManager handling the commands, I've changed so that logic now parses the commands, and checks whether user is authenticated to execute certain command.

One big problem though is that, how AddressBookParser works is that it first displays the invalid message command before it actually parse the arguments, but I would want to check authentication first before the invalid message occurs. Look at picture below. This is before the user is actually logged in.

image

As you can see even if the command is not fully typed, it should display this error message as seen below 

image

### CODE REVIEW

**Comment by nianfei97** [comment]
1. Check whether the format is corrent/Try to match command word
    1a. If wrong format/command doesn't exist (Parse exception), print invalid command message
2. Check authentication
    2a. If not authorised (Authorize exception), print unauthorised message
3. Try to parse individual command
    3a. If cannot parse (Parse exception), print usage message
4. Try to execute
    4a. If cannot execute (Command exception), print usage message

Feels like the key is in splitting up point 1 and point 3 to insert point 2 in between. Now `parseCommand()` in `AddressBookParser` does point 1 and 3 in the same method, that's why it's causing this problem. Maybe consider splitting `parseCommand()` into, say `identifyCommandWord()`, which returns the correct command class (e.g. `AddTimeCommand.class`) and `parseCommandArgs()`, which returns the corresponding command if successfully parsed.

**Comment by nianfei97** [comment]
Maybe `parseCommand()` can return a `Class` object (https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html) so security knows what command it is. Then check for authentication. Then try to parse it by calling the correct parser (somehow link the Class object to the corresponding parser)

**Comment by bennchong** [comment]
Pulled upstream before comitting, fixed insane number of merge conflicts. Please do merge if passes all tests

**Comment by nianfei97** [comment]
Follow-up question: Even if we ignore the SRP-related "problems", why does checking for security need to happen twice (once during parsing and once during executing)? Seems to be double work

File: docs/LearningOutcomes.adoc [hunk 1]
```diff
@@ -190,7 +190,7 @@ public void execute_invalidCommandFormat_throwsParseException() {
 
 == Write System Tests `[LO-SystemTesting]`
 
-Note how tests below `src/test/java/systemtests` package (e.g link:{repoURL}/src/test/java/systemtests/AddCommandSystemTest.java[`AddCommandSystemTest.java`]) are system tests because they test the entire system end-to-end.
+Note how tests below `src/test/java/systemtests` package (e.g link:{repoURL}/src/test/java/systemtests/RegisterCommandSystemTest.java[`RegisterCommandSystemTest.java`]) are system tests because they test the entire system end-to-end.
```
[the remark this answers is not in the document]
Comment (bennchong) [reply docs/LearningOutcomes.adoc]: Lol just edited it to be correct

File: src/main/java/seedu/address/commons/events/security/GetAuthenticationReplyEvent.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package seedu.address.commons.events.security;
+
+import seedu.address.commons.events.BaseEvent;
+
+/** Indicates the AddressBook in the model has changed*/
+public class GetAuthenticationReplyEvent extends BaseEvent {
```
[the remark this answers is not in the document]
Comment (bennchong) [reply src/main/java/seedu/address/commons/events/security/GetAuthenticationReplyEvent.java]: Hmm true ill change this

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 3]
```diff
@@ -30,6 +30,7 @@ public FindCommand(NameContainsKeywordsPredicate predicate) {
     @Override
     public CommandResult execute(Model model, CommandHistory history) {
```
[the remark this answers is not in the document]
Comment (bennchong) [reply src/main/java/seedu/address/logic/commands/FindCommand.java]: No it throws it in the Parser

File: src/main/java/seedu/address/logic/commands/TagCommand.java [hunk 4]
```diff
@@ -31,6 +31,7 @@ public TagCommand(TagsContainsKeywords tagsKeywords) {
     @Override
     public CommandResult execute(Model model, CommandHistory history) throws CommandException {
```
[the remark this answers is not in the document]
Comment (bennchong) [reply src/main/java/seedu/address/logic/commands/TagCommand.java]: Same^

File: src/main/java/seedu/address/logic/parser/AddTimeCommandParser.java [hunk 5]
```diff
@@ -2,26 +2,42 @@
...
+
+    private boolean isAuthenticated;
 
     /**
      * Parses the given {@code String} of arguments in the context of the AddTimeCommand
      * and returns an {@code AddTimeCommand} object for execution.
      * @throws ParseException if the user input does not conform the expected format
      */
-    public AddTimeCommand parse (String args) throws ParseException {
+    public AddTimeCommand parse (String args) throws ParseException, SecurityAuthenticationException {
```
[the remark this answers is not in the document]
Comment (bennchong) [reply src/main/java/seedu/address/logic/parser/AddTimeCommandParser.java]: Cause all the parser checks for Security

File: src/main/java/seedu/address/logic/parser/ParserClass.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package seedu.address.logic.parser;
+
+import seedu.address.commons.core.EventsCenter;
+import seedu.address.commons.events.BaseEvent;
+
+/**
+ * Represents a Parser Class with EventsCenter
+ */
+public abstract class ParserClass {
```
[the remark this answers is not in the document]
Comment (bennchong) [reply src/main/java/seedu/address/logic/parser/ParserClass.java]: Yea the Parser is required to have an EventsBus associated with them

File: src/main/java/seedu/address/model/person/Address.java [hunk 7]
```diff
@@ -34,6 +34,9 @@ public Address(String address) {
      * Returns true if a given string is a valid address.
      */
     public static boolean isValidAddress(String test) {
+        if (test.equals("")) {
```
[the remark this answers is not in the document]
Comment (bennchong) [reply src/main/java/seedu/address/model/person/Address.java]: Cause if you want the address to be empty to be valid this ensures it

**Comment by bennchong** [comment]
> Follow-up question: Even if we ignore the SRP-related "problems", why does checking for security need to happen twice (once during parsing and once during executing)? Seems to be double work

It doesn't happen during executing, only parsing

File: src/main/java/seedu/address/logic/commands/ClearCommand.java [hunk 8]
```diff
@@ -18,7 +19,7 @@
     public static final String MESSAGE_SUCCESS = "Your timetable has been cleared!";
 
     @Override
-    public CommandResult execute(Model model, CommandHistory history) {
+    public CommandResult execute(Model model, CommandHistory history) throws SecurityAuthenticationException {
```
Comment (nianfei97) [inline src/main/java/seedu/address/logic/commands/ClearCommand.java]: If it's the parser that throws the security exception, why does this throw an exception too?

**Comment by bennchong** [comment]
> One big question:
> Why do the parsers throw security exceptions? Seems to be a big violation of SRP

Good question. Lets take it on slack to explain

File: src/main/java/seedu/address/logic/commands/ClearCommand.java (hunk 8 above)
Comment (bennchong) [reply src/main/java/seedu/address/logic/commands/ClearCommand.java]: Cause for this commands that do not have parsers, this is the place to put

**Comment by bennchong** [comment]
Pretty sure the issues will drop by 5 once this cleared

**Comment by bennchong** [comment]
To @nianfei97 questions above, perhaps the reply on slack will be hidden, ill copy paste here.

1. You are right it doesn't need, my new commits removes it.

2. Yep we need logic for keeping track of history, which is declared in logic

**Pull request merged**